package com.kh.rupp_dev.boukryuniversity.controller;


import com.kh.rupp_dev.boukryuniversity.payload.MultipleResponse;
import com.kh.rupp_dev.boukryuniversity.payload.SingleResponse;
import com.kh.rupp_dev.boukryuniversity.dto.request.PaginationRequest;
import org.springframework.data.domain.Page;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import com.kh.rupp_dev.boukryuniversity.dto.request.ClassRequest;
import com.kh.rupp_dev.boukryuniversity.dto.response.ClassResponse;
import com.kh.rupp_dev.boukryuniversity.service.ClassService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequiredArgsConstructor
@RequestMapping("/classes")
public class ClassController {

	private final ClassService classService;

	@PostMapping
	public ResponseEntity<SingleResponse<ClassResponse>> create(@Valid @RequestBody ClassRequest classRequest) {
		ClassResponse response = classService.create(classRequest);
		return ResponseEntity.ok(SingleResponse.success("Successfully created class.", response));
	}

	@GetMapping
	public ResponseEntity<MultipleResponse<ClassResponse>> getAll(PaginationRequest request) {
		Page<ClassResponse> responses = classService.getAll(request.toPageable());
		return ResponseEntity
				.ok(MultipleResponse.success("Successfully retrieved all classes with pagination.", responses));
	}

	@GetMapping("/{id}")
	public ResponseEntity<SingleResponse<ClassResponse>> getById(@PathVariable Long id) {
		ClassResponse response = classService.getById(id);
		return ResponseEntity.ok(SingleResponse.success("Successfully retrieved class.", response));
	}

	@PutMapping("/{id}")
	public ResponseEntity<SingleResponse<ClassResponse>> update(@PathVariable Long id,
			@RequestBody @Valid ClassRequest classRequest) {
		ClassResponse response = classService.update(id, classRequest);
		return ResponseEntity.ok(SingleResponse.success("Successfully updated class.", response));
	}

	@DeleteMapping("/{id}")
	public ResponseEntity<SingleResponse<ClassResponse>> deleteById(@PathVariable Long id) {
		classService.delete(id);
		return ResponseEntity.ok(SingleResponse.success("Successfully deleted class.", null));
	}
}
